server/core/src/main/java/org/infinispan/server/core/security/ServerAuthenticationProvider.java

```diff
@@ -11,18 +21,75 @@
-    *           the SASL mechanism to get a callback handler for
-    * @param mechanismProperties
-    *           the mechanism properties that might need to be adjusted to support the specific mechanism / callbackhandler combination
+    * @param mechanismName       the SASL mechanism to get a callback handler for
+    * @param mechanismProperties the mechanism properties that might need to be adjusted to support the specific
+    *                            mechanism / callbackhandler combination
     * @return the callback handler or {@code null} if the mechanism is not supported
+    * @deprecated This method will no longer be invoked directly. Implement the
+    * {@link #createSaslServer(String, List, String, String, Map)} method instead.
+    */
+   @Deprecated
+   default AuthorizingCallbackHandler getCallbackHandler(String mechanismName, Map<String, String> mechanismProperties) {
+      return null;
+   }
+
+   /**
+    * Create a SaslServer, to be used for a single authentication session, for the specified mechanismName. On
+    * completion of the SASL authentication exchange, the SaslServer <b>MUST</b> provide a non-read-only negotiated
+    * {@link javax.security.auth.Subject} when {@link SaslServer#getNegotiatedProperty(String)} is invoked with the
+    * {@link SubjectSaslServer#SUBJECT} property. The default implementation of this method will
+    *
+    * @param mechanism  The non-null IANA-registered name of a SASL mechanism. (e.g. "GSSAPI", "CRAM-MD5").
+    * @param principals Any principals which can be obtained before the authentication (e.g. TLS peer, remote network
+    *                   address). Can be empty.
+    * @param protocol   The non-null string name of the protocol for which the authentication is being performed (e.g.,
+    *                   "ldap").
+    * @param serverName The fully qualified host name of the server to authenticate to, or null if the server is not
+    *                   bound to any specific host name. If the mechanism does not allow an unbound server, a {@code
+    *                   SaslException} will be thrown.
+    * @param props      The possibly null set of properties used to select the SASL mechanism and to configure the
+    *                   authentication exchange of the selected mechanism. See the {@code Sasl} class for a list of
+    *                   standard properties. Other, possibly mechanism-specific, properties can be included. Properties
+    *                   not relevant to the selected mechanism are ignored, including any map entries with non-String
+    *                   keys.
+    * @return an instance of SaslServer (or null if it cannot be created)
     */
-   AuthorizingCallbackHandler getCallbackHandler(String mechanismName, Map<String, String> mechanismProperties);
+   default SaslServer createSaslServer(String mechanism, List<Principal> principals, String protocol, String serverName, Map<String, String> props) throws SaslException {
```

Comment:
Should we add a `ClassLoader` param here so that we only have to implement `getCallbackHandler` in `EndpointServerAuthenticationProvider`?

Reply:
Used this.getClass().getClassLoader()